### ISSUE TITLE
[Feature] Skip steps via CLI arg

### ISSUE THREAD

**Issue by mmkal** [microsoft/playwright#18690]
Would be quite cool to do this:

```ts
test('first-time user flow', async ({page}) => {
  const user = await test.step('signup', async () => {
    await page.goto(url + '/signup')
    const user = await someHelper.createTestUser()
    await page.locator('.username').type(user.username)
    await page.locator('.password').type(user.password)
    // lots more steps, taking quite a long time
    return user
  })

  await test.step('dashboard', async () => {
    await page.goto(url + '/account')
    await page.click('.messages')
    await page.locator(`text=Welcome, ${user.username}!`).waitFor()
  })
})
```

Then, let's say I set some sort of imaginary config in playwright.config.ts:

```ts
  stepsOutput: 'dist/steps',
```

This could write a file, say for test run with id `abc123`, it'd write something like this to `dist/steps/abc123.json`:

```json
[
  {
    "file": "my-file.spec.ts",
    "test": "first-time user flow",
    "steps": [
      "signup",
      {
        "username": "xyz",
        "password": "qwertyuiop"
      }
    ]
  }
]
```

Then when re-running, instead of doing `yarn test my-file.spec.ts`, you could do `yarn test my-file.spec.ts  --restore-steps abc123`. That would tell playwright to look for `dist/steps/abc123.json`, parse it, and match the `test` and `steps` it finds, to _skip_ `await test.step` using the values stored in there. That way, a developer can fix problems in the _last_ steps (either in the product code, or the test assertions), and skip ahead to the part they're debugging when re-running tests.

This could also be done in userland, but we'd have to wrap `test` and/or `test.step`. It'd be great if playwright could create some official conventions around it.

Also, looking at the above, it might be possible to read a json output file. Not sure if step outputs are serialised in there. I'm assuming not, because I'm not sure `test.step` outputs have to be serialisable in the first place?

**Comment by mmkal** [comment]
Yes, that sounds like more or less it. The one add-on it'd be good to include, though, is to get it "for free" just by using `test.step`. Since we could use the (e.g. `signup` in `test.step('signup', ...)` as the key and the return value of the step's function as the value. Ideally it wouldn't be necessary to add any lines of code to get this speed up for free.

Worth noting, I have a hacky version working locally here:

```ts
import {promises as fs} from 'fs'
import * as path from 'path'
import {test as base} from '@playwright/test'

export const test = base.extend<{
  step: typeof base.step
}>({
  step: async ({}, use, testInfo) => {
    await use(async <T>(title: string, fn: () => Promise<T>): Promise<T> => {
      if (!process.env.STEP_HYDRATION) {
        return base.step(title, fn)
      }

      const slugify = (s: string) => s.replace(/\W+/g, '-')
      const filepath = path.join(
        process.cwd(),
        'dist/test-steps', // don't use testInfo.outputDir or testInfo.outputPath(...) because it's wiped for every new run
        ...testInfo.titlePath.map(slugify),
        slugify(title) + '.json',
      )

      console.log('reading', {filepath, previous})
      const content = await fs.readFile(filepath).catch(() => '{}')
      const {output: previous} = JSON.parse(content.toString()) as {output?: T}

      if (previous) {
        return base.step(title, async () => previous)
      }

      const output = await base.step(title, fn)

      console.log('writing', {filepath, output})
      await fs.mkdir(path.dirname(filepath), {recursive: true})
      await fs.writeFile(filepath, JSON.stringify({output}))

      return output
    })
  },
})
```

Can be used like this:

```ts
test('first-time user flow', async ({page, step}) => {
  const user = await step('signup', async () => {
    await page.goto(url + '/signup')
    const user = await someHelper.createTestUser()
    await page.locator('.username').type(user.username)
    await page.locator('.password').type(user.password)
    // lots more steps, taking quite a long time
    return user
  })

  await step('dashboard', async () => {
    await page.goto(url + '/account')
    await page.click('.messages')
    await page.locator(`text=Welcome, ${user.username}!`).waitFor()
  })
})
```

So basically, the injected `step` works exactly the same way as `test.step`, but if you set the magic `STEP_HYDRATION` environment variable, intermediate steps that have already been run will be skipped very quickly, passing the same data as last time to future steps. It works, but it's hacky and not documented I think it'd be better as part of playwright, especially if it fits into existing key/value cache plans!

**Comment by yury-s** [comment]
> Yes, that sounds like more or less it. 

We postponed that API release until later. I'll keep this bug open for now and close it once we land the storage API, so that you get notified. You can then give it a try and file a new issue with the feedback based on that API experience if anything is still not working.

**Comment by mmkal** [comment]
Thanks!

**Comment by yury-s** [comment]
We've tried a couple of approaches with key value store that could be used to cache the results between tests but were not satisfied with the added value of any of those and decided not to add them to playwright for now. So for now you best bet is to use the`step` fixture to implement this functionality. Looks like you have everything in the current API for implementing it as a fixture, please let us know if that's not the case.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
